lib/elixir/src/elixir_clauses.erl

```diff
@@ -40,8 +40,28 @@ guard({'when', Meta, [Left, Right]}, E) ->
   {ELeft, EL}  = guard(Left, E),
   {ERight, ER} = guard(Right, EL),
   {{'when', Meta, [ELeft, ERight]}, ER};
-guard(Other, E) ->
-  elixir_expand:expand(Other, E).
+guard(Guard, E) ->
+  {EGuard, EG} = elixir_expand:expand(Guard, E),
+  ok = warn_zero_length_guard(EGuard, EG),
+  {EGuard, EG}.
+
+warn_zero_length_guard(
+  {
+    {'.', _, [erlang, '==']},
+    Meta,
+    [{{'.', _, [erlang, length]}, _, [Arg]}, 0]
+  },
```

Comment:
I had to change it, since the expanded `Kernel.length` is inlined to `:erlang.length`.